### ISSUE TITLE
Broken balance-slb failover and performance issues with VLANs & bonds

### ISSUE THREAD

**Issue by Manouchehri** [openvswitch/ovs-issues#136]
First, here's the full description of my setup that works  as expected (>1Gbps) without OVS: https://blogger.davidmanouchehri.com/2017/09/fast-ipsec-on-embedded-devices-with.html

Now, I'm trying to implement this with OVS, but I'm getting terrible results.

![Network](https://2.bp.blogspot.com/-m3_C1MI7MPY/Waz0Ttz1HdI/AAAAAAAAIeE/YSTXW__GaYkQxlfCSB0xjtS1o5L9EIy0gCLcBGAs/s1600/unnamed0.png)

Summary of problems:
- All outgoing traffic is stuck  on one VLAN and one enpXs0 port
- Failover doesn't work at all for the VLANs. (Only works for enp1s0/enp2s0/enp3s0/enp4s0.)
- The first outgoing VLAN can be fully saturated with multiple connections, and the second VLAN is still never even touched.

<details>
  <summary>Setup + Logs (click to open)</summary>

# Setup
```bash
dave@intel:~$ sudo ovs-vsctl \
	-- add-br ovs-physical \
	-- add-bond ovs-physical intel enp1s0 enp2s0 enp3s0 enp4s0 bond_mode=balance-slb \
	-- add-port ovs-physical vlan2372 tag=2372 \
	-- add-port ovs-physical vlan2662 tag=2662 \
	-- set interface vlan2372 type=patch \
	-- set interface vlan2662 type=patch \
	-- set interface vlan2372 options:peer=vvlan2372 \
	-- set interface vlan2662 options:peer=vvlan2662 \
	-- add-br ovs-wan \
	-- add-bond ovs-wan vwan vvlan2372 vvlan2662 bond_mode=balance-slb \
	-- set interface vvlan2372 type=patch \
	-- set interface vvlan2662 type=patch \
	-- set interface vvlan2372 options:peer=vlan2372 \
	-- set interface vvlan2662 options:peer=vlan2662 \
	-- set bridge ovs-physical other-config:hwaddr=00:20:91:CD:BA:00 \
	-- set bridge ovs-wan other-config:hwaddr=00:20:91:FC:FE:00
```
# `ovs-appctl bond/show`
```bash
dave@intel:~$ sudo ovs-appctl bond/show
---- intel ----
bond_mode: balance-slb
bond may use recirculation: no, Recirc-ID : -1
bond-hash-basis: 0
updelay: 0 ms
downdelay: 0 ms
next rebalance: 1385 ms
lacp_status: off
active slave mac: 00:0e:c4:d0:5d:96(enp2s0)

slave enp1s0: enabled
        may_enable: true

slave enp2s0: enabled
        active slave
        may_enable: true
        hash 46: 11 kB load

slave enp3s0: enabled
        may_enable: true

slave enp4s0: enabled
        may_enable: true

---- vwan ----
bond_mode: balance-slb
bond may use recirculation: no, Recirc-ID : -1
bond-hash-basis: 0
updelay: 0 ms
downdelay: 0 ms
next rebalance: 1385 ms
lacp_status: off
active slave mac: 6e:bf:b5:f1:23:f4(vvlan2662)

slave vvlan2372: enabled
        may_enable: true

slave vvlan2662: enabled
        active slave
        may_enable: true
        hash 126: 11 kB load
```

# `ovs-vsctl show`
```bash
dave@intel:~$ sudo ovs-vsctl show
e56273f3-2bbe-4524-81c3-2af3646ef9a4
    Bridge ovs-wan
        Port ovs-wan
            Interface ovs-wan
                type: internal
        Port vwan
            Interface "vvlan2662"
                type: patch
                options: {peer="vlan2662"}
            Interface "vvlan2372"
                type: patch
                options: {peer="vlan2372"}
    Bridge ovs-physical
        Port intel
            Interface "enp4s0"
            Interface "enp1s0"
            Interface "enp3s0"
            Interface "enp2s0"
        Port "vlan2372"
            tag: 2372
            Interface "vlan2372"
                type: patch
                options: {peer="vvlan2372"}
        Port ovs-physical
            Interface ovs-physical
                type: internal
        Port "vlan2662"
            tag: 2662
            Interface "vlan2662"
                type: patch
                options: {peer="vvlan2662"}
    ovs_version: "2.5.2"
```

# `ovs-vsctl list interface`
```bash
_uuid               : d4f66cb9-5c63-4e5b-af9d-5cd11c543eab
admin_state         : up
bfd                 : {}
bfd_status          : {}
cfm_fault           : []
cfm_fault_status    : []
cfm_flap_count      : []
cfm_health          : []
cfm_mpid            : []
cfm_remote_mpids    : []
cfm_remote_opstate  : []
duplex              : full
error               : []
external_ids        : {}
ifindex             : 3
ingress_policing_burst: 0
ingress_policing_rate: 0
lacp_current        : []
link_resets         : 1
link_speed          : 1000000000
link_state          : up
lldp                : {}
mac                 : []
mac_in_use          : "00:0e:c4:d0:5d:96"
mtu                 : 1500
name                : "enp2s0"
ofport              : 5
ofport_request      : []
options             : {}
other_config        : {}
statistics          : {collisions=0, rx_bytes=194655438155, rx_crc_err=0, rx_dropped=0, rx_errors=0, rx_frame_err=0, rx_over_err=0, rx_packets=130299978, tx_bytes=28402500376, tx_dropped=0, tx_errors=0, tx_packets=77357767}
status              : {driver_name=igb, driver_version="5.3.0-k", firmware_version=" 0. 6-1"}
type                : ""

_uuid               : 00a217b1-e4fa-4f6b-850a-e0e983915c2a
admin_state         : up
bfd                 : {}
bfd_status          : {}
cfm_fault           : []
cfm_fault_status    : []
cfm_flap_count      : []
cfm_health          : []
cfm_mpid            : []
cfm_remote_mpids    : []
cfm_remote_opstate  : []
duplex              : []
error               : []
external_ids        : {}
ifindex             : 9
ingress_policing_burst: 0
ingress_policing_rate: 0
lacp_current        : []
link_resets         : 1
link_speed          : []
link_state          : up
lldp                : {}
mac                 : []
mac_in_use          : "00:20:91:fc:fe:00"
mtu                 : 1500
name                : ovs-wan
ofport              : 65534
ofport_request      : []
options             : {}
other_config        : {}
statistics          : {collisions=0, rx_bytes=192309504502, rx_crc_err=0, rx_dropped=116913, rx_errors=0, rx_frame_err=0, rx_over_err=0, rx_packets=130292615, tx_bytes=28093070609, tx_dropped=0, tx_errors=0, tx_packets=77357773}
status              : {driver_name=openvswitch}
type                : internal

_uuid               : 1c597e31-7b1d-4e35-b27b-032f6dbee790
admin_state         : up
bfd                 : {}
bfd_status          : {}
cfm_fault           : []
cfm_fault_status    : []
cfm_flap_count      : []
cfm_health          : []
cfm_mpid            : []
cfm_remote_mpids    : []
cfm_remote_opstate  : []
duplex              : full
error               : []
external_ids        : {}
ifindex             : 5
ingress_policing_burst: 0
ingress_policing_rate: 0
lacp_current        : []
link_resets         : 1
link_speed          : 1000000000
link_state          : up
lldp                : {}
mac                 : []
mac_in_use          : "00:0e:c4:d0:5d:98"
mtu                 : 1500
name                : "enp4s0"
ofport              : 2
ofport_request      : []
options             : {}
other_config        : {}
statistics          : {collisions=0, rx_bytes=1190010, rx_crc_err=0, rx_dropped=0, rx_errors=0, rx_frame_err=0, rx_over_err=0, rx_packets=12964, tx_bytes=818, tx_dropped=0, tx_errors=0, tx_packets=7}
status              : {driver_name=igb, driver_version="5.3.0-k", firmware_version=" 0. 6-1"}
type                : ""

_uuid               : 53c7bbba-4845-4846-99c7-2b0256433b3f
admin_state         : up
bfd                 : {}
bfd_status          : {}
cfm_fault           : []
cfm_fault_status    : []
cfm_flap_count      : []
cfm_health          : []
cfm_mpid            : []
cfm_remote_mpids    : []
cfm_remote_opstate  : []
duplex              : []
error               : []
external_ids        : {}
ifindex             : 0
ingress_policing_burst: 0
ingress_policing_rate: 0
lacp_current        : []
link_resets         : 0
link_speed          : []
link_state          : up
lldp                : {}
mac                 : []
mac_in_use          : "1a:4d:b7:42:d2:0a"
mtu                 : []
name                : "vlan2372"
ofport              : 6
ofport_request      : []
options             : {peer="vvlan2372"}
other_config        : {}
statistics          : {collisions=0, rx_bytes=0, rx_crc_err=0, rx_dropped=0, rx_errors=0, rx_frame_err=0, rx_over_err=0, rx_packets=0, tx_bytes=330011, tx_dropped=0, tx_errors=0, tx_packets=2544}
status              : {}
type                : patch

_uuid               : 92c88eda-2ede-4092-ab84-c5812edbdb95
admin_state         : up
bfd                 : {}
bfd_status          : {}
cfm_fault           : []
cfm_fault_status    : []
cfm_flap_count      : []
cfm_health          : []
cfm_mpid            : []
cfm_remote_mpids    : []
cfm_remote_opstate  : []
duplex              : full
error               : []
external_ids        : {}
ifindex             : 4
ingress_policing_burst: 0
ingress_policing_rate: 0
lacp_current        : []
link_resets         : 1
link_speed          : 1000000000
link_state          : up
lldp                : {}
mac                 : []
mac_in_use          : "00:0e:c4:d0:5d:97"
mtu                 : 1500
name                : "enp3s0"
ofport              : 4
ofport_request      : []
options             : {}
other_config        : {}
statistics          : {collisions=0, rx_bytes=1189946, rx_crc_err=0, rx_dropped=0, rx_errors=0, rx_frame_err=0, rx_over_err=0, rx_packets=12963, tx_bytes=818, tx_dropped=0, tx_errors=0, tx_packets=7}
status              : {driver_name=igb, driver_version="5.3.0-k", firmware_version=" 0. 6-1"}
type                : ""

_uuid               : 4cbbbd4b-a8a1-43a5-a07c-fce43124e395
admin_state         : up
bfd                 : {}
bfd_status          : {}
cfm_fault           : []
cfm_fault_status    : []
cfm_flap_count      : []
cfm_health          : []
cfm_mpid            : []
cfm_remote_mpids    : []
cfm_remote_opstate  : []
duplex              : []
error               : []
external_ids        : {}
ifindex             : 0
ingress_policing_burst: 0
ingress_policing_rate: 0
lacp_current        : []
link_resets         : 0
link_speed          : []
link_state          : up
lldp                : {}
mac                 : []
mac_in_use          : "d2:52:1f:7e:f0:46"
mtu                 : []
name                : "vlan2662"
ofport              : 1
ofport_request      : []
options             : {peer="vvlan2662"}
other_config        : {}
statistics          : {collisions=0, rx_bytes=28093070468, rx_crc_err=0, rx_dropped=0, rx_errors=0, rx_frame_err=0, rx_over_err=0, rx_packets=77357772, tx_bytes=194654759817, tx_dropped=0, tx_errors=0, tx_packets=130292478}
status              : {}
type                : patch

_uuid               : 87393e84-8650-4ec7-be26-7711dac2b42f
admin_state         : up
bfd                 : {}
bfd_status          : {}
cfm_fault           : []
cfm_fault_status    : []
cfm_flap_count      : []
cfm_health          : []
cfm_mpid            : []
cfm_remote_mpids    : []
cfm_remote_opstate  : []
duplex              : []
error               : []
external_ids        : {}
ifindex             : 0
ingress_policing_burst: 0
ingress_policing_rate: 0
lacp_current        : []
link_resets         : 0
link_speed          : []
link_state          : up
lldp                : {}
mac                 : []
mac_in_use          : "6e:bf:b5:f1:23:f4"
mtu                 : []
name                : "vvlan2662"
ofport              : 1
ofport_request      : []
options             : {peer="vlan2662"}
other_config        : {}
statistics          : {collisions=0, rx_bytes=194654759817, rx_crc_err=0, rx_dropped=0, rx_errors=0, rx_frame_err=0, rx_over_err=0, rx_packets=130292478, tx_bytes=28093070468, tx_dropped=0, tx_errors=0, tx_packets=77357772}
status              : {}
type                : patch

_uuid               : 218f7aee-b035-406c-ab7d-14c15216817f
admin_state         : up
bfd                 : {}
bfd_status          : {}
cfm_fault           : []
cfm_fault_status    : []
cfm_flap_count      : []
cfm_health          : []
cfm_mpid            : []
cfm_remote_mpids    : []
cfm_remote_opstate  : []
duplex              : full
error               : []
external_ids        : {}
ifindex             : 2
ingress_policing_burst: 0
ingress_policing_rate: 0
lacp_current        : []
link_resets         : 1
link_speed          : 1000000000
link_state          : up
lldp                : {}
mac                 : []
mac_in_use          : "00:0e:c4:d0:5d:95"
mtu                 : 1500
name                : "enp1s0"
ofport              : 3
ofport_request      : []
options             : {}
other_config        : {}
statistics          : {collisions=0, rx_bytes=1189946, rx_crc_err=0, rx_dropped=0, rx_errors=0, rx_frame_err=0, rx_over_err=0, rx_packets=12963, tx_bytes=818, tx_dropped=0, tx_errors=0, tx_packets=7}
status              : {driver_name=igb, driver_version="5.3.0-k", firmware_version=" 0. 6-1"}
type                : ""

_uuid               : ac2b88d6-275d-428d-ad60-e05aaec5444e
admin_state         : up
bfd                 : {}
bfd_status          : {}
cfm_fault           : []
cfm_fault_status    : []
cfm_flap_count      : []
cfm_health          : []
cfm_mpid            : []
cfm_remote_mpids    : []
cfm_remote_opstate  : []
duplex              : []
error               : []
external_ids        : {}
ifindex             : 0
ingress_policing_burst: 0
ingress_policing_rate: 0
lacp_current        : []
link_resets         : 0
link_speed          : []
link_state          : up
lldp                : {}
mac                 : []
mac_in_use          : "c6:60:00:82:ab:3e"
mtu                 : []
name                : "vvlan2372"
ofport              : 2
ofport_request      : []
options             : {peer="vlan2372"}
other_config        : {}
statistics          : {collisions=0, rx_bytes=330011, rx_crc_err=0, rx_dropped=0, rx_errors=0, rx_frame_err=0, rx_over_err=0, rx_packets=2544, tx_bytes=0, tx_dropped=0, tx_errors=0, tx_packets=0}
status              : {}
type                : patch

_uuid               : c7bd070f-5363-43a0-9fb8-d481229b11f0
admin_state         : up
bfd                 : {}
bfd_status          : {}
cfm_fault           : []
cfm_fault_status    : []
cfm_flap_count      : []
cfm_health          : []
cfm_mpid            : []
cfm_remote_mpids    : []
cfm_remote_opstate  : []
duplex              : []
error               : []
external_ids        : {}
ifindex             : 8
ingress_policing_burst: 0
ingress_policing_rate: 0
lacp_current        : []
link_resets         : 1
link_speed          : []
link_state          : up
lldp                : {}
mac                 : []
mac_in_use          : "00:20:91:cd:ba:00"
mtu                 : 1500
name                : ovs-physical
ofport              : 65534
ofport_request      : []
options             : {}
other_config        : {}
statistics          : {collisions=0, rx_bytes=848350, rx_crc_err=0, rx_dropped=16, rx_errors=0, rx_frame_err=0, rx_over_err=0, rx_packets=11037, tx_bytes=738, tx_dropped=0, tx_errors=0, tx_packets=7}
status              : {driver_name=openvswitch}
type                : internal
```

# `ovs-vsctl list port`

```bash
_uuid               : e9e3fa78-53e8-48ba-b18b-ec7aea405172
bond_active_slave   : []
bond_downdelay      : 0
bond_fake_iface     : false
bond_mode           : []
bond_updelay        : 0
external_ids        : {}
fake_bridge         : false
interfaces          : [4cbbbd4b-a8a1-43a5-a07c-fce43124e395]
lacp                : []
mac                 : []
name                : "vlan2662"
other_config        : {}
qos                 : []
rstp_statistics     : {}
rstp_status         : {}
statistics          : {}
status              : {}
tag                 : 2662
trunks              : []
vlan_mode           : []

_uuid               : bd3e1bed-dfbe-4197-b562-47ce7c51b5ad
bond_active_slave   : []
bond_downdelay      : 0
bond_fake_iface     : false
bond_mode           : []
bond_updelay        : 0
external_ids        : {}
fake_bridge         : false
interfaces          : [c7bd070f-5363-43a0-9fb8-d481229b11f0]
lacp                : []
mac                 : []
name                : ovs-physical
other_config        : {}
qos                 : []
rstp_statistics     : {}
rstp_status         : {}
statistics          : {}
status              : {}
tag                 : []
trunks              : []
vlan_mode           : []

_uuid               : 99edfd09-13d5-47f1-95fb-2d36edc6d284
bond_active_slave   : []
bond_downdelay      : 0
bond_fake_iface     : false
bond_mode           : []
bond_updelay        : 0
external_ids        : {}
fake_bridge         : false
interfaces          : [53c7bbba-4845-4846-99c7-2b0256433b3f]
lacp                : []
mac                 : []
name                : "vlan2372"
other_config        : {}
qos                 : []
rstp_statistics     : {}
rstp_status         : {}
statistics          : {}
status              : {}
tag                 : 2372
trunks              : []
vlan_mode           : []

_uuid               : 09b93d86-67ed-4194-80c9-ff9e7e61932b
bond_active_slave   : []
bond_downdelay      : 0
bond_fake_iface     : false
bond_mode           : []
bond_updelay        : 0
external_ids        : {}
fake_bridge         : false
interfaces          : [00a217b1-e4fa-4f6b-850a-e0e983915c2a]
lacp                : []
mac                 : []
name                : ovs-wan
other_config        : {}
qos                 : []
rstp_statistics     : {}
rstp_status         : {}
statistics          : {}
status              : {}
tag                 : []
trunks              : []
vlan_mode           : []

_uuid               : 1afca74e-d402-4110-b909-b745826d6cb4
bond_active_slave   : "00:0e:c4:d0:5d:96"
bond_downdelay      : 0
bond_fake_iface     : false
bond_mode           : balance-slb
bond_updelay        : 0
external_ids        : {}
fake_bridge         : false
interfaces          : [1c597e31-7b1d-4e35-b27b-032f6dbee790, 218f7aee-b035-406c-ab7d-14c15216817f, 92c88eda-2ede-4092-ab84-c5812edbdb95, d4f66cb9-5c63-4e5b-af9d-5cd11c543eab]
lacp                : []
mac                 : []
name                : intel
other_config        : {}
qos                 : []
rstp_statistics     : {}
rstp_status         : {}
statistics          : {}
status              : {}
tag                 : []
trunks              : []
vlan_mode           : []

_uuid               : 3819eedb-8af6-4a8d-a099-af06bf6139be
bond_active_slave   : "6e:bf:b5:f1:23:f4"
bond_downdelay      : 0
bond_fake_iface     : false
bond_mode           : balance-slb
bond_updelay        : 0
external_ids        : {}
fake_bridge         : false
interfaces          : [87393e84-8650-4ec7-be26-7711dac2b42f, ac2b88d6-275d-428d-ad60-e05aaec5444e]
lacp                : []
mac                 : []
name                : vwan
other_config        : {}
qos                 : []
rstp_statistics     : {}
rstp_status         : {}
statistics          : {}
status              : {}
tag                 : []
trunks              : []
vlan_mode           : []
```

# `iperf3`

```bash
dave@intel:~$ iperf3 -c iperf.he.net --zerocopy --parallel 8 --reverse --omit 30 -t 60 --format m --interval 60 -4
Connecting to host iperf.he.net, port 5201
Reverse mode, remote host iperf.he.net is sending
[  4] local 192.168.0.19 port 42752 connected to 216.218.227.10 port 5201
[  6] local 192.168.0.19 port 42754 connected to 216.218.227.10 port 5201
[  8] local 192.168.0.19 port 42756 connected to 216.218.227.10 port 5201
[ 10] local 192.168.0.19 port 42758 connected to 216.218.227.10 port 5201
[ 12] local 192.168.0.19 port 42760 connected to 216.218.227.10 port 5201
[ 14] local 192.168.0.19 port 42762 connected to 216.218.227.10 port 5201
[ 16] local 192.168.0.19 port 42764 connected to 216.218.227.10 port 5201
[ 18] local 192.168.0.19 port 42766 connected to 216.218.227.10 port 5201
[ ID] Interval           Transfer     Bandwidth
[  4]   0.00-60.00  sec   780 MBytes   109 Mbits/sec
[  6]   0.00-60.00  sec   603 MBytes  84.3 Mbits/sec
[  8]   0.00-60.00  sec   574 MBytes  80.2 Mbits/sec
[ 10]   0.00-60.00  sec   897 MBytes   125 Mbits/sec
[ 12]   0.00-60.00  sec   738 MBytes   103 Mbits/sec
[ 14]   0.00-60.00  sec   551 MBytes  77.0 Mbits/sec
[ 16]   0.00-60.00  sec   659 MBytes  92.2 Mbits/sec
[ 18]   0.00-60.00  sec   729 MBytes   102 Mbits/sec
[SUM]   0.00-60.00  sec  5.40 GBytes   773 Mbits/sec
- - - - - - - - - - - - - - - - - - - - - - - - -
[ ID] Interval           Transfer     Bandwidth       Retr
[  4]   0.00-60.00  sec   782 MBytes   109 Mbits/sec    9             sender
[  4]   0.00-60.00  sec   781 MBytes   109 Mbits/sec                  receiver
[  6]   0.00-60.00  sec   604 MBytes  84.4 Mbits/sec    8             sender
[  6]   0.00-60.00  sec   604 MBytes  84.4 Mbits/sec                  receiver
[  8]   0.00-60.00  sec   575 MBytes  80.4 Mbits/sec    8             sender
[  8]   0.00-60.00  sec   574 MBytes  80.3 Mbits/sec                  receiver
[ 10]   0.00-60.00  sec   899 MBytes   126 Mbits/sec    9             sender
[ 10]   0.00-60.00  sec   898 MBytes   126 Mbits/sec                  receiver
[ 12]   0.00-60.00  sec   739 MBytes   103 Mbits/sec    8             sender
[ 12]   0.00-60.00  sec   739 MBytes   103 Mbits/sec                  receiver
[ 14]   0.00-60.00  sec   552 MBytes  77.2 Mbits/sec    8             sender
[ 14]   0.00-60.00  sec   552 MBytes  77.1 Mbits/sec                  receiver
[ 16]   0.00-60.00  sec   660 MBytes  92.3 Mbits/sec   10             sender
[ 16]   0.00-60.00  sec   660 MBytes  92.3 Mbits/sec                  receiver
[ 18]   0.00-60.00  sec   731 MBytes   102 Mbits/sec    9             sender
[ 18]   0.00-60.00  sec   730 MBytes   102 Mbits/sec                  receiver
[SUM]   0.00-60.00  sec  5.41 GBytes   775 Mbits/sec   69             sender
[SUM]   0.00-60.00  sec  5.41 GBytes   774 Mbits/sec                  receiver

iperf Done.
```

</details>

### tl;dr: OVS gets close to half the speed compared to my "standard" network. What am I doing wrong here?
